### PR TITLE
Make OpenGL the default on macOS

### LINKED ISSUE #62: Segfault when scrolling through a dataset
When loading and plotting one of my test BDF files, I get a segfault when scrolling a little through the data. Here's the code I type into IPython:

```python
import mne

raw = mne.io.read_raw("S01.bdf")
raw.plot(block=True)
```

And the full output after the plotting command:

```
Using qt as 2D backend.
Opening raw-browser...
Free RAM space could not be determined because"psutil" is not installed. Setting precompute to False.
qt.qpa.fonts: Populating font family aliases took 367 ms. Replace uses of missing font family "AnyStyle" with one that exists to avoid this cost. 
Fatal Python error: Segmentation fault

Thread 0x000070000d54d000 (most recent call first):
  File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 320 in wait
  File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 600 in wait
  File "/usr/local/lib/python3.10/site-packages/IPython/core/history.py", line 829 in run
  File "/usr/local/lib/python3.10/site-packages/IPython/core/history.py", line 60 in only_when_enabled
  File "/usr/local/lib/python3.10/site-packages/decorator.py", line 232 in fun
  File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 1009 in _bootstrap_inner
  File "/usr/local/Cellar/python@3.10/3.10.2/Frameworks/Python.framework/Versions/3.10/lib/python3.10/threading.py", line 966 in _bootstrap

Current thread 0x0000000116715e00 (most recent call first):
  File "/usr/local/lib/python3.10/site-packages/pyqtgraph/graphicsItems/PlotCurveItem.py", line 563 in paint
  File "/usr/local/lib/python3.10/site-packages/pyqtgraph/debug.py", line 110 in w
  File "/usr/local/lib/python3.10/site-packages/pyqtgraph/widgets/GraphicsView.py", line 143 in paintEvent
  File "/Users/clemens/Repositories/mne-python/mne/viz/backends/_utils.py", line 164 in _qt_app_exec
  File "/Users/clemens/Repositories/mne-python/mne/viz/utils.py", line 153 in _show_browser
  File "/Users/clemens/Repositories/mne-python/mne/viz/_figure.py", line 652 in _get_browser
  File "/Users/clemens/Repositories/mne-python/mne/viz/raw.py", line 353 in plot_raw
  File "<decorator-gen-180>", line 12 in plot_raw
  File "/Users/clemens/Repositories/mne-python/mne/io/base.py", line 1551 in plot
  File "<ipython-input-6-2fe91485cf70>", line 1 in <module>
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3251 in run_code
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3191 in run_ast_nodes
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3012 in run_cell_async
  File "/usr/local/lib/python3.10/site-packages/IPython/core/async_helpers.py", line 129 in _pseudo_sync_runner
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2814 in _run_cell
  File "/usr/local/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2768 in run_cell
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/interactiveshell.py", line 604 in interact
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/interactiveshell.py", line 611 in mainloop
  File "/usr/local/lib/python3.10/site-packages/IPython/terminal/ipapp.py", line 316 in start
  File "/usr/local/lib/python3.10/site-packages/traitlets/config/application.py", line 846 in launch_instance
  File "/usr/local/lib/python3.10/site-packages/IPython/__init__.py", line 123 in start_ipython
  File "/usr/local/bin/ipython", line 8 in <module>

Extension modules: black.const, blib2to3.pgen2.token, blib2to3.pgen2.grammar, blib2to3.pytree, blib2to3.pgen2.tokenize, blib2to3.pgen2.parse, blib2to3.pgen2.pgen, blib2to3.pgen2.driver, blib2to3.pygram, black.mode, black.cache, black.strings, black.nodes, black.brackets, black.lines, black.comments, black.numerics, black.rusty, black.trans, black.linegen, black.handle_ipynb_magics, black.parsing, black, numpy.core._multiarray_umath, numpy.core._multiarray_tests, numpy.linalg._umath_linalg, numpy.fft._pocketfft_internal, numpy.random._common, numpy.random.bit_generator, numpy.random._bounded_integers, numpy.random._mt19937, numpy.random.mtrand, numpy.random._philox, numpy.random._pcg64, numpy.random._sfc64, numpy.random._generator, numpy.linalg.lapack_lite, PyQt5.QtCore, PyQt5.QtGui, PyQt5.QtWidgets, PyQt5.QtTest, matplotlib._c_internal_utils, PIL._imaging, matplotlib._path, kiwisolver, matplotlib._image, PyQt5.QtSvg, scipy._lib._ccallback_c, scipy.sparse._sparsetools, scipy.sparse._csparsetools, scipy.sparse.csgraph._tools, scipy.sparse.csgraph._shortest_path, scipy.sparse.csgraph._traversal, scipy.sparse.csgraph._min_spanning_tree, scipy.sparse.csgraph._flow, scipy.sparse.csgraph._matching, scipy.sparse.csgraph._reordering, scipy.spatial._ckdtree, scipy._lib.messagestream, scipy.spatial._qhull, scipy.spatial._voronoi, scipy.linalg._fblas, scipy.linalg._flapack, scipy.linalg._cythonized_array_utils, scipy.linalg._flinalg, scipy.linalg._solve_toeplitz, scipy.linalg._matfuncs_sqrtm_triu, scipy.linalg.cython_blas, scipy.linalg.cython_lapack, scipy.linalg._decomp_update, scipy.spatial._distance_wrap, scipy.spatial._hausdorff, scipy.special._ufuncs_cxx, scipy.special._ufuncs, scipy.special._specfun, scipy.special._comb, scipy.special._ellip_harm_2, scipy.spatial.transform._rotation, scipy.ndimage._nd_image, _ni_label, scipy.ndimage._ni_label, scipy.sparse.linalg._isolve._iterative, scipy.sparse.linalg._dsolve._superlu, scipy.sparse.linalg._eigen.arpack._arpack, scipy.optimize._minpack2, scipy.optimize._group_columns, scipy.optimize._trlib._trlib, scipy.optimize._lbfgsb, _moduleTNC, scipy.optimize._moduleTNC, scipy.optimize._cobyla, scipy.optimize._slsqp, scipy.optimize._minpack, scipy.optimize._lsq.givens_elimination, scipy.optimize._zeros, scipy.optimize.__nnls, scipy.optimize._highs.cython.src._highs_wrapper, scipy.optimize._highs._highs_wrapper, scipy.optimize._highs.cython.src._highs_constants, scipy.optimize._highs._highs_constants, scipy.linalg._interpolative, scipy.optimize._bglu_dense, scipy.optimize._lsap_module, scipy.integrate._odepack, scipy.integrate._quadpack, scipy.integrate._vode, scipy.integrate._dop, scipy.integrate._lsoda, scipy.interpolate._fitpack, scipy.interpolate.dfitpack, scipy.interpolate._bspl, scipy.interpolate._ppoly, scipy.interpolate.interpnd, scipy.interpolate._rbfinterp_pythran, scipy.special.cython_special, scipy.stats._stats, beta_ufunc, scipy.stats._boost.beta_ufunc, binom_ufunc, scipy.stats._boost.binom_ufunc, nbinom_ufunc, scipy.stats._boost.nbinom_ufunc, hypergeom_ufunc, scipy.stats._boost.hypergeom_ufunc, scipy.stats._biasedurn, scipy.stats._hypotests_pythran, scipy.stats._statlib, scipy.stats._mvn, scipy.stats._sobol, scipy.stats._qmc_cy, scipy.stats._unuran.unuran_wrapper, objc._objc, CoreFoundation._inlines, CoreFoundation._CoreFoundation, Foundation._Foundation, Foundation._inlines (total: 136)
[1]    55701 segmentation fault  ipython
```

If it's due to some property of the file I'm using (which works fine with the Matplotlib backend), I can try to share it privately (it's pretty large though).

```
Platform:         macOS-11.6.4-x86_64-i386-64bit
Python:           3.10.2 (main, Feb  2 2022, 08:42:42) [Clang 13.0.0 (clang-1300.0.29.3)]
Executable:       /usr/local/opt/python@3.10/bin/python3.10
CPU:              i386: 8 cores
Memory:           Unavailable (requires "psutil" package)
mne:              1.0.dev0
numpy:            1.22.2 {}
scipy:            1.8.0
matplotlib:       3.5.1 {backend=MacOSX}

sklearn:          1.0.2
numba:            Not found
nibabel:          Not found
nilearn:          Not found
dipy:             Not found
cupy:             Not found
pandas:           1.4.1
pyvista:          Not found
pyvistaqt:        Not found
ipyvtklink:       Not found
vtk:              Not found
PyQt5:            5.15.6
ipympl:           Not found
pooch:            v1.6.0

mne_bids:         Not found
mne_nirs:         Not found
mne_features:     Not found
mne_qt_browser:   0.2.1
mne_connectivity: Not found
```

### PR DESCRIPTION
**Pull request by cbrnr**
Fixes #62, or at least works around it.

### CODE REVIEW

File: mne_qt_browser/_pg_figure.py [hunk 1]
```diff
@@ -2780,22 +2780,20 @@ def __init__(self, **kwargs):
...
-                     'Consider installing pyopengl with pip or conda'
-                     'or set "use_opengl" to False to avoid this warning.')
-                if mac_epochs:
-                    warn('Plotting epochs on MacOS without OpenGL'
-                         'may be unstable!')
+                warn('PyOpenGL was not found and OpenGL cannot be used. '
+                     'Consider installing pyopengl with pip or conda or set '
+                     '"use_opengl=False" to avoid this warning.')
+                if sys.platform == 'darwin':
+                    warn('Plotting on macOS without OpenGL may be unstable!')
```
Comment (hoechenberger) [inline mne_qt_browser/_pg_figure.py:2796]: I think we should raise an exception honestly, if we cannot even guarantee we're not going to crash!

Comment (cbrnr) [reply mne_qt_browser/_pg_figure.py:2796]: Fine with me. Which exception would you raise?

Comment (hoechenberger) [reply mne_qt_browser/_pg_figure.py:2796]: re-raise the `ImportError`, maybe together with the warning text you have here

Comment (hoechenberger) [reply mne_qt_browser/_pg_figure.py:2796]: Hm ok in that case, a clear warning might be okay. I'm just worried it might first get unnoticed and later become invisible because the segfault output fills up the scrollback buffer

Comment (larsoner) [reply mne_qt_browser/_pg_figure.py:2796]: https://github.com/mne-tools/mne-qt-browser#development-and-testing says

> pytest -m benchmark mne_qt_browser

For me, it's faster for raw, slower for epochs, and sometimes causes errors. So I'd just leave it off on my system...

One option is to raise if the value is not passed (i.e., if `use_opengl=None` to start) *and* it's not set in the config var MNE_BROWSER_USE_OPENGL. That would make it so that a naive user will see an error, but an expert user who explicitly passes `use_opengl=False` or sets MNE_BROWSER_USE_OPENGL=false will not see an error. In these expert cases I actually wouldn't even bother to emit a warning.

Comment (cbrnr) [reply mne_qt_browser/_pg_figure.py:2796]: `E   FileNotFoundError: fname does not exist: /Users/clemens/Repositories/mne-qt-browser/.venv/lib/python3.10/site-packages/mne/io/tests/data/test_raw.fif`

How do I get this file?

And yes, I like your suggestion!

Comment (larsoner) [reply mne_qt_browser/_pg_figure.py:2796]: If you `pip install -e .` or `python setup.py develop` your locally cloned MNE-Python it'll be there. We don't ship the testing files with the PyPi release, but they're there in the GitHub repo.

Comment (cbrnr) [reply mne_qt_browser/_pg_figure.py:2796]: So I need to clone `mne-python` inside `mne-qt-browser`? Can we put `pip install -U -e git+https://github.com/mne-tools/mne-python.git#egg=mne` in the development docs for future reference (in case it's not mentioned already)?

Comment (cbrnr) [reply mne_qt_browser/_pg_figure.py:2796]: There is absolutely *no* difference between `use_opengl=True` and `use_opengl=False` on my Mac.

**Comment by larsoner** [comment]
> So I need to clone mne-python inside mne-qt-browser?

No, you just need to have it cloned *somewhere* and installed in editable mode.

> Can we put pip install -U -e git+https://github.com/mne-tools/mne-python.git#egg=mne in the development docs for future reference (in case it's not mentioned already)?

I would just tell people to use a standard editable / developer install of MNE-Python and leave it at that. The assumption is that anyone working on mne-qt-browser will probably also have (or know how to get) such an install set up

File: mne_qt_browser/_pg_figure.py (hunk 1 above)
Comment (cbrnr) [reply mne_qt_browser/_pg_figure.py:2796]: So about that behavior: I would like to make `use_opengl=True` the default on macOS. We warn if `pyopengl` is not installed and set `use_opengl=False`, but this case is unlikely to happen when we include `pyopengl` as a requirement on macOS.

In all other cases, it seems like at least it is beneficial when plotting epochs. But I really don't want to make any changes here that I haven't tested. I haven't used this browser at all, so I would like you to decide how to proceed.

**Comment by cbrnr** [comment]
> No, you just need to have it cloned somewhere and installed in editable mode.

How does this work? I have it installed in some other separate venv.

**Comment by larsoner** [comment]
To clarify, I mean you should have it cloned somewhere and installed in editable mode *in the same python (v or conda)env where you are developing/editing mne-qt-browser*. So you could clone it again and install in editable mode if you want, or if both venvs have the same Python version (and maybe even if they're different?) you can `pip install -e.` and `python setup.py develop` to multiple envs at once.

File: mne_qt_browser/_pg_figure.py (hunk 1 above)
Comment (larsoner) [reply mne_qt_browser/_pg_figure.py:2796]: > I would like you to decide how to proceed.

My vote is for my proposal in https://github.com/mne-tools/mne-qt-browser/pull/63#issuecomment-1051052302 . Maybe it's just easier for me to push a commit to do it?

**Comment by cbrnr** [comment]
OK, sounds still a bit complicated. `pip install -e .` will only work when I'm in the `mne-python` repo with the way I set up my terminal (I'm using `autoenv`), so I'd prefer `pip install`ing it again directly from GitHub. In any case, I don't think that this is super clear to everyone wanting to contribute (or running the tests at least), so maybe we could be more specific and mention a few options how to install MNE-Python?

File: mne_qt_browser/_pg_figure.py (hunk 1 above)
Comment (cbrnr) [reply mne_qt_browser/_pg_figure.py:2796]: Yes, please go ahead and make the changes!

**Comment by cbrnr** [comment]
And TBH, if all of this effort is just to get the test data, wouldn't it be easier to download those files from their GitHub URL?

**Comment by larsoner** [comment]
> And TBH, if all of this effort is just to get the test data, wouldn't it be easier to download those files from their GitHub URL?

I think for most devs of mne-qt-browser there is little to no extra effort involved, because wherever they are developing / editing mne-qt-browser, they are also developing / editing mne-python. Your setup with `autoenv` seems like a corner case in this sense, so I'd rather not add additional complications to the testing infrastructure or instructions for it.

I would just add to the dev instructions something like "To run tests, you'll need an editable (not pip or conda) MNE-Python install in order for the necessary test files to be available". This doesn't seem onerous to me? But I guess it's okay to add one line of an example of how to do it using `pip`, as well, it's easy enough to ignore for people who have simpler setups.

**Comment by cbrnr** [comment]
> I would just add to the dev instructions something like "To run tests, you'll need an editable (not pip or conda) MNE-Python install in order for the necessary test files to be available". This doesn't seem onerous to me? But I guess it's okay to add one line of an example of how to do it using pip, as well, it's easy enough to ignore for people who have simpler setups.

Sounds good! Yes, please add the `pip install -e git+https:...` example, I find it easier than the other options. If in the future more people find this procedure too complicated we can think about alternatives. I mean, if we need only one or two data files I'd just download them.

**Review by marsipu — APPROVED** [review]

**Comment by larsoner** [comment]
> Yes, please add the pip install -e git+https:... example, I find it easier than the other options

Feel free to do this in another PR. I think we should get this one in and release 0.2.2 ASAP so that it can be rolled into the 1.0 installer. The dev instructions can be updated whenever since people will probably go to GitHub to see them anyway!
